File: SeriesGuide/src/main/java/com/battlelancer/seriesguide/ui/search/EpisodeSearchFragment.java
package com.battlelancer.seriesguide.ui.search;

import android.app.SearchManager;
import android.content.Intent;
import android.database.Cursor;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.LoaderManager.LoaderCallbacks;
import android.support.v4.content.CursorLoader;
import android.support.v4.content.Loader;
import android.view.View;
import android.widget.AdapterView;
import com.battlelancer.seriesguide.provider.SeriesGuideContract.EpisodeSearch;
import com.battlelancer.seriesguide.provider.SeriesGuideContract.Shows;
import com.battlelancer.seriesguide.provider.SeriesGuideDatabase;
import com.battlelancer.seriesguide.ui.SearchActivity;
import com.battlelancer.seriesguide.ui.episodes.EpisodesActivity;
import com.battlelancer.seriesguide.util.TabClickEvent;
import com.battlelancer.seriesguide.util.Utils;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;

/**
 * Displays episode search results.
 */
public class EpisodeSearchFragment extends BaseSearchFragment {

    private EpisodeResultsAdapter adapter;

    public interface InitBundle {
        /** Set to pre-filter search results by show title. */
        String SHOW_TITLE = "title";
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        // list items do not have right hand-side buttons, list may be long: enable fast scrolling
        gridView.setFastScrollAlwaysVisible(false);
        gridView.setFastScrollEnabled(true);
    }

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        adapter = new EpisodeResultsAdapter(getActivity());
        gridView.setAdapter(adapter);
    }

    @Override
    public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
        Intent i = new Intent(getActivity(), EpisodesActivity.class);
        i.putExtra(EpisodesActivity.InitBundle.EPISODE_TVDBID, (int) id);

        Utils.startActivityWithAnimation(getActivity(), i, view);
    }

    @Subscribe(sticky = true, threadMode = ThreadMode.MAIN)
    public void onEventMainThread(SearchActivity.SearchQueryEvent event) {
        search(event.args);
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onEventTabClick(TabClickEvent event) {
        if (event.position == SearchActivity.TAB_POSITION_EPISODES) {
            gridView.smoothScrollToPosition(0);
        }
    }

    private void search(Bundle args) {
        getLoaderManager().restartLoader(SearchActivity.EPISODES_LOADER_ID, args,
                searchLoaderCallbacks);
    }

    private LoaderCallbacks<Cursor> searchLoaderCallbacks = new LoaderCallbacks<Cursor>() {
        @Override
        public Loader<Cursor> onCreateLoader(int id, Bundle args) {
            String selection = null;
            final String query = args.getString(SearchManager.QUERY);
            String[] selectionArgs = new String[] {
                    query
            };

            Bundle appData = args.getBundle(SearchManager.APP_DATA);
            if (appData != null) {
                String showtitle = appData.getString(InitBundle.SHOW_TITLE);
                // set show filter instead
                if (showtitle != null) {
                    selection = Shows.TITLE + "=?";
                    selectionArgs = new String[] {
                            query, showtitle
                    };
                }
            }

            return new CursorLoader(getActivity(), EpisodeSearch.CONTENT_URI_SEARCH,
                    SeriesGuideDatabase.EpisodeSearchQuery.PROJECTION, selection, selectionArgs,
                    null);
        }

        @Override
        public void onLoadFinished(Loader<Cursor> loader, Cursor data) {
            adapter.swapCursor(data);
        }

        @Override
        public void onLoaderReset(Loader<Cursor> loader) {
            adapter.swapCursor(null);
        }
    };
}

File: app/src/main/java/com/battlelancer/seriesguide/shows/search/EpisodeSearchFragment.kt
package com.battlelancer.seriesguide.shows.search

import android.app.SearchManager
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.GridView
import androidx.fragment.app.viewModels
import com.battlelancer.seriesguide.R
import com.battlelancer.seriesguide.databinding.FragmentSearchBinding
import com.battlelancer.seriesguide.shows.episodes.EpisodesActivity
import com.battlelancer.seriesguide.util.TabClickEvent
import com.battlelancer.seriesguide.util.Utils
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode

/**
 * Displays episode search results.
 */
class EpisodeSearchFragment : BaseSearchFragment() {

    private var binding: FragmentSearchBinding? = null
    private val model by viewModels<EpisodeSearchViewModel>()
    private lateinit var adapter: EpisodeSearchAdapter

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        return FragmentSearchBinding.inflate(inflater, container, false)
            .also { binding = it }
            .root
    }

    override val emptyView: View
        get() = binding!!.textViewSearchEpisodesEmpty
    override val gridView: GridView
        get() = binding!!.gridViewSearchEpisodes

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        // list items do not have right hand-side buttons, list may be long: enable fast scrolling
        gridView.apply {
            isFastScrollAlwaysVisible = false
            isFastScrollEnabled = true
        }
        adapter = EpisodeSearchAdapter(requireContext(), onItemClickListener).also {
            gridView.adapter = it
        }

        model.episodes.observe(viewLifecycleOwner) { episodes ->
            adapter.setData(episodes)
            updateEmptyState(
                episodes.isEmpty(),
                !model.searchData.value?.searchTerm.isNullOrEmpty()
            )
        }

        // load for given query (if just created)
        val args = initialSearchArgs
        if (args != null) {
            updateQuery(args)
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        binding = null
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onEventMainThread(event: SearchActivityImpl.SearchQueryEvent) {
        updateQuery(event.args)
    }

    private fun updateQuery(args: Bundle) {
        model.searchData.value = EpisodeSearchViewModel.SearchData(
            args.getString(SearchManager.QUERY),
            args.getBundle(SearchManager.APP_DATA)?.getString(ARG_SHOW_TITLE)
        )
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onEventTabClick(event: TabClickEvent) {
        if (event.position == SearchActivityImpl.TAB_POSITION_EPISODES) {
            gridView.smoothScrollToPosition(0)
        }
    }

    private val onItemClickListener = object : EpisodeSearchAdapter.OnItemClickListener {
        override fun onItemClick(anchor: View, episodeId: Long) {
            EpisodesActivity.intentEpisode(episodeId, requireContext())
                .also { Utils.startActivityWithAnimation(activity, it, anchor) }
        }
    }

    companion object {
        val liftOnScrollTargetViewId = R.id.gridViewSearchEpisodes

        const val ARG_SHOW_TITLE = "title"
    }
}
